Kernel/Bus/USB/UHCI/UHCIController.cpp

```diff
@@ -510,6 +519,41 @@ ErrorOr<size_t> UHCIController::submit_bulk_transfer(Transfer& transfer)
     return transfer_size;
 }
 
+ErrorOr<void> UHCIController::submit_async_interrupt_transfer(NonnullLockRefPtr<Transfer> transfer, u16 ms_interval)
+{
+    dbgln_if(UHCI_DEBUG, "UHCI: Received interrupt transfer for address {}. Root Hub is at address {}.", transfer->pipe().device_address(), m_root_hub->device_address());
+
+    if (ms_interval == 0) {
+        return EINVAL;
+    }
+
+    SpinlockLocker locker(m_async_lock);
+
+    auto transfer_queue = TRY(create_transfer_queue(*transfer));
+
+    u8 idx = 0;
+    while ((1 << idx) < ms_interval && idx < NUMBER_OF_INTERRUPT_QHS) {
+        idx++;
+    }
+
+    m_active_async_transfers.append({ transfer, transfer_queue, ms_interval });
+    enqueue_qh(transfer_queue, m_interrupt_qh_anchor_arr[idx]);
+
+    return {};
+}
+
+void UHCIController::cancel_async_transfer(NonnullLockRefPtr<Transfer> transfer)
+{
+    SpinlockLocker async_locker(m_async_lock);
+    for (size_t i = 0; i < m_active_async_transfers.size(); i++) {
+        if (m_active_async_transfers[i].transfer.ptr() == transfer.ptr()) {
+            dequeue_qh(m_active_async_transfers[i].qh);
+            m_active_async_transfers.remove(i);
```

Comment:
Ah something else. You are removing an item from this list while iterating it. Which... is "ok" since you return right after doing so, but generally feels awkward. This is also a `find_if` algorithm. Refactoring to use AK::find_if (or a similar member on vector? I don't remember if there is one) would make the intent here clearer.

Reply:
Or even remove_if, if we have one.